--- ohphone/Makefile.orig	2003-07-13 18:32:31.000000000 +0200
+++ ohphone/Makefile	2003-07-13 18:33:21.000000000 +0200
@@ -151,10 +151,10 @@
 SOURCES		:= main.cxx vidlinux.cxx xlibvid.cxx sdlvid.cxx
 
 ifndef OPENH323DIR
-OPENH323DIR=$(HOME)/openh323
+OPENH323DIR=/usr
 endif
 
-include $(OPENH323DIR)/openh323u.mak
+include /usr/share/openh323/openh323u.mak
 
 #
 # only include svgalib support if available on this sytem
