diff -durN ohphone.orig/Makefile ohphone/Makefile
--- ohphone.orig/Makefile	Sat Jan  6 12:23:48 2001
+++ ohphone/Makefile	Sat Jan  6 12:24:32 2001
@@ -117,14 +117,6 @@
 
 include $(OPENH323DIR)/share/misc/openh323u.mak
 
-#
-# only include svgalib support if available on this sytem
-#
-ifneq (,$(wildcard /usr/include/vga.h))
-LDLIBS		+= -lvga
-STDCCFLAGS	+= -DHAS_VGALIB
-endif
-
 ifdef	HAS_X11
 LDFLAGS     	+= -L/usr/X11R6/lib
 LDLIBS		+= -lXext -lX11
