--- ohphone/Makefile.orig	Tue Jul  9 18:18:33 2002
+++ ohphone/Makefile	Tue Jul  9 18:20:22 2002
@@ -152,11 +152,11 @@
 #
 # only include svgalib support if available on this sytem
 #
-ifneq (,$(wildcard $(SYSINCDIR)/vga.h))
-LDLIBS		+= -lvga
-STDCCFLAGS	+= -DHAS_VGALIB
-endif
-
+#ifneq (,$(wildcard $(SYSINCDIR)/vga.h))
+#LDLIBS		+= -lvga
+#STDCCFLAGS	+= -DHAS_VGALIB
+#endif
+#
 #
 # only include X Windows support if available on this sytem
 #
